linkerd/trace-context/src/propagation/w3c.rs

```diff
@@ -0,0 +1,148 @@
+use http::HeaderValue;
+use tracing::{trace, warn};
+
+use super::{decode_id_with_padding, get_header_str, Propagation, TraceContext};
+use crate::{Flags, Id};
+
+pub const HTTP_TRACEPARENT: &str = "traceparent";
+pub const HEADER_VERSION_00: &str = "00";
+
+pub fn unpack_w3c_trace_context<B>(request: &http::Request<B>) -> Option<TraceContext> {
+    get_header_str(request, HTTP_TRACEPARENT).and_then(parse_context)
+}
+
+/// Given an http request and a w3c trace context, create a new Span ID and
+/// assign it to the tracecontext header value, in order to propagate
+/// the trace context downstream.
+pub fn increment_http_span_id<B>(request: &mut http::Request<B>, context: &TraceContext) -> Id {
+    let span_id = Id::new_span_id(&mut rand::thread_rng());
+
+    trace!("incremented span id: {}", span_id);
+
+    let new_header = {
+        let mut buf = String::with_capacity(60);
+        buf.push_str(HEADER_VERSION_00);
+        buf.push('-');
+        buf.push_str(&hex::encode(context.trace_id.as_ref()));
+        buf.push('-');
+        buf.push_str(&hex::encode(context.parent_id.as_ref()));
+        buf.push('-');
+        buf.push_str(&hex::encode(vec![context.flags.0]));
+        buf
+    };
+
+    if let Result::Ok(hv) = HeaderValue::from_str(&new_header) {
+        request.headers_mut().insert(HTTP_TRACEPARENT, hv);
+    } else {
+        warn!("invalid value {} for tracecontext header", new_header);
+    }
```

Comment:
Compiler will complain because we're not return an Option from the second branch. I wonder if that's why we used the `if let` syntax instead of a match? Do you know if we can somehow ignore the return types in the match arms?

Reply:
Also, had to change everything to be `static` and we're passing references now. Compiler yells otherwise for moving out of static item (which I guess makes lots of sense).